src/Kafka/Producer.hs

```diff
@@ -108,6 +114,62 @@ produceMessage kp@(KafkaProducer (Kafka k) _ (TopicConf tc)) m = liftIO $ do
                 keyPtr (fromIntegral keyLength) nullPtr
 
 
+-- | Sends a single message synchronously
+--
+--   This function implements a simple synchronous production of a message by
+--   passing along an error pointer. Once this pointer has been set to an error
+--   the message can be considered sent. If the message is sent with
+--   'RdKafkaRespErrNoError', then this returns a `Nothing`
+--
+--   Example C-solution used: https://github.com/edenhill/librdkafka/wiki/Sync-producer
+produceMessageSync :: MonadIO m
+                   => KafkaProducer
+                   -> ProducerRecord
+                   -> m (Maybe KafkaError)
+produceMessageSync (KafkaProducer (Kafka k) _ (TopicConf tc)) m = liftIO $ do
+  bracket (mkTopic $ prTopic m) clTopic withTopic
+    where
+      mkTopic (TopicName tn) = newUnmanagedRdKafkaTopicT k (Text.unpack tn) (Just tc)
+
+      clTopic = either (return . const ()) destroyUnmanagedRdKafkaTopic
+
+      withTopic (Left err) = pure . Just . KafkaError $ Text.pack err
+      withTopic (Right t) =
+        withBS (prValue m) $ \payloadPtr payloadLength ->
+          withBS (prKey m) $ \keyPtr keyLength ->
+            withPtr $ \resPtr ->
+              let
+                produce =
+                  rdKafkaProduce t (producePartitionCInt (prPartition m))
+                    copyMsgFlags payloadPtr (fromIntegral payloadLength)
+                    keyPtr (fromIntegral keyLength) resPtr
+              in do
+                res <- produce
+                if res == (-1 :: Int) then
+                  (Just . kafkaRespErr) <$> getErrno
+                else
+                  waitForAck resPtr
+
+      initialError :: Int
+      initialError = -12345
+
+      withPtr :: (Ptr RdKafkaRespErrT -> IO (Maybe KafkaError)) -> IO (Maybe KafkaError)
+      withPtr f = alloca $ \ptr -> do
+        poke ptr initialError
+        f $ castPtr ptr
+
+      waitForAck :: Ptr RdKafkaRespErrT -> IO (Maybe KafkaError)
+      waitForAck ptr = do
+        currentVal <- peek $ castPtr ptr
+        if currentVal == initialError then do
+          _ <- rdKafkaPoll k 1000
```

Comment:
In the presence of an async `poll` loop (which `hw-kafka-client` runs), this may imply an unnecessary delay.
Imagine that the async loop got there first and delivered the callbacks. This means that this loop will wait for the end of the second before it checks the value again and exits.
Which makes sync delivery incredibly slow.